java-checks/src/main/java/org/sonar/java/checks/MethodParametersOrderCheck.java

```diff
@@ -0,0 +1,121 @@
+    MethodInvocationTree methodInvTree = (MethodInvocationTree) tree;
+    List<Symbol> definedFunctionParameters = parameterNamesByMethod.get(methodInvTree.symbol());
+    MethodTree methodTree = (MethodTree) methodInvTree.symbol().declaration();
+    if (definedFunctionParameters == null) {
+      if (methodTree != null) {
+        definedFunctionParameters = methodTree.parameters().stream().map(VariableTree::symbol).collect(Collectors.toCollection(ArrayList::new));
+        parameterNamesByMethod.put(methodTree.symbol(), definedFunctionParameters);
+      } else {
+        return;
+      }
+    }
+    List<IdentifierTree> argumentsList = methodInvTree.arguments().stream().map(arg -> {
+      if (arg.is(Tree.Kind.IDENTIFIER)) {
+        return ((IdentifierTree) arg);
+      } else if (arg.is(Tree.Kind.MEMBER_SELECT)) {
+        return (IdentifierTree) ExpressionUtils.skipParentheses(((MemberSelectExpressionTree) arg).identifier());
+      } else {
+        return null;
+      }
+    }).collect(Collectors.toList());
+    if (argumentsNamesMatchParametersNames(definedFunctionParameters, argumentsList)
+      && argumentsTypesMatchParametersTypesAndNamesNotOrdered(definedFunctionParameters, argumentsList)) {
+      List<JavaFileScannerContext.Location> flow = methodTree.parameters().stream().map(param -> new JavaFileScannerContext.Location("Formal Parameters", param))
+        .collect(Collectors.toList());
+      reportIssue(methodInvTree.arguments(), "Parameters to " + methodInvTree.symbol().name() + " have the same names but not the same order as the method arguments.",
+        flow, null);
+
+    }
+  }
+
+  private static boolean argumentsNamesMatchParametersNames(List<Symbol> parametersList, List<IdentifierTree> argumentsList) {
+    List<String> argListNames = argumentsList.stream().filter(Objects::nonNull).map(parameter -> parameter.name().toLowerCase(Locale.ENGLISH)).collect(Collectors.toList());
+    return allUnique(argListNames)
+      && argListNames.stream().allMatch(arg -> parametersList.stream().map(parameter -> parameter.name().toLowerCase(Locale.ENGLISH)).collect(Collectors.toList()).contains(arg));
+
+  }
+
+  public static boolean allUnique(List<String> argListNames) {
+    return argListNames.stream().allMatch(new HashSet<>()::add);
+  }
+
+  private static boolean argumentsTypesMatchParametersTypesAndNamesNotOrdered(List<Symbol> parameterList, List<IdentifierTree> argumentList) {
+    Iterator<IdentifierTree> argumentsIterator = argumentList.stream().filter(Objects::nonNull).iterator();
+    int countArgumentsNotOrdered = 0;
+    while (argumentsIterator.hasNext()) {
+      IdentifierTree argument = argumentsIterator.next();
+      int index = (parameterList.stream().map(param -> param.name().toLowerCase(Locale.ENGLISH)).collect(Collectors.toList()))
+        .indexOf(argument.name().toLowerCase(Locale.ENGLISH));
+      if (parameterList.get(index).type().equals(argument.symbolType())) {
```

Comment:
we also need to deal with unknown types. If the formal type is unknown or the argument type is unknown, we should return false.